File: flutter17/iconfont.py
#-*- coding:utf-8 -*-
import re
import os

# 根据 Iconton.cn下载的图标文件自动生成 /lib/util/iconfont.dart文件
iconCss = "./assets/fonts/iconfont.css"
iconName = 'iconfont'
path = './lib/util/'
dartPath = './lib/util/iconfont.dart'

fileStr = """import 'package:flutter/cupertino.dart';

// 文件由iconfont.py自动生成

class MyIcon {
--
}
"""


def main():
    with open(iconCss) as f:
        str = f.read()
        str = str.replace("\n  content", "content")
        
        result = []
        code = []
        for row in str.split('\n'):
            m = re.match(r'.icon-(.*?):.*\\(.*)";', row)
            if m != None:
                name, val = m.group(1).replace("-", "_"), m.group(2)
                result.append((name, val))
        
        for (name, val) in result:
            str = "  static const IconData %s = const IconData(0x%s, fontFamily: '%s');" %(name, val, iconName)
            code.append(str)

        if not os.path.exists(path):
            os.makedirs(path)
        with open(dartPath, "w+") as fs:
            fileContent = fileStr.replace("--", "\n".join(code))
            fs.write(fileContent)
    
if __name__ == '__main__':
    main()

File: flutter17/iconfont.js
const fs = require('fs')
const path = require('path')

/**
 *  从 Iconfont.cn下载图表后将所有文件放在 /assets/fonts下，执行 node iconfont.js
 *  读取 assets/iconfont.css, 解析内容, 生成 /lib/iconfont.dart
 */

const IconfontCss = path.resolve(__dirname, 'assets/fonts/iconfont.css')
const fontName = 'iconfont'
const dartPath = path.resolve(__dirname, 'lib/iconfont.dart')

let data = fs.readFileSync(IconfontCss).toString()

while (data.includes('\n  content')) {
    data = data.replace('\n  content', 'content')
}

// .*? 表示非贪婪搜索
const reg = /\.icon-(.*?):.*\\(.*)";/

let result = []
data.split('\n').forEach(row => {
    let r = reg.exec(row)
    if (r) {
        result.push({
            name: r[1].replace('-', '_'),
            value: r[2]
        })
    }
})


let dartFile = `
import 'package:flutter/cupertino.dart';

class MyIcon {
--
}
`
let code = [], string = ''
result.forEach(item => {
    code.push(`  static const IconData ${item.name} = const IconData(0x${item.value}, fontFamily: '${fontName}');`)
})

string = code.join('\n')

dartFile = dartFile.replace('--', string)

fs.writeFileSync(dartPath, dartFile)
